### PR TITLE
feat(next-core): support unsupported module runtime error

### PR DESCRIPTION
**Pull request by kwonoj**
### What

Implement webpack's middleware plugin equivalent for Turbopack, to raise unsupported error in runtime. 

PR utilizes import map alias for the edge context, to resolve into modulereplacer internally provides a virtualsource to call runtime error logic. Since we already have globally augmented, the virtualsource only need to take export those into module.

Closes PACK-2789

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: b3990ffd638ae87434a2db31a59ec58ad6f723c1

`TURBOPACK=1 pnpm test test/integration/edge-runtime-module-errors/test/index.test.js` (turbopack)
- Edge runtime code with imports > Edge API using Buffer polyfill > does not throw in dev at runtime
- Edge runtime code with imports > Middleware using Buffer polyfill > does not throw in dev at runtime

<details>
<summary>Expand output</summary>

  ● Edge runtime code with imports › Edge API using Buffer polyfill › does not throw in dev at runtime

    expect(received).toBe(expected) // Object.is equality

    Expected: 200
    Received: 500

      389 |       context.app = await launchApp(context.appDir, context.appPort, appOption)
      390 |       const res = await fetchViaHTTP(context.appPort, url)
    > 391 |       expect(res.status).toBe(200)
          |                          ^
      392 |       expect(res.headers.get('x-from-runtime')).toBe('false')
      393 |       expectNoError(moduleName)
      394 |     })

      at Object.toBe (integration/edge-runtime-module-errors/test/index.test.js:391:26)

  ● Edge runtime code with imports › Middleware using Buffer polyfill › does not throw in dev at runtime

    expect(received).toBe(expected) // Object.is equality

    Expected: 200
    Received: 500

      389 |       context.app = await launchApp(context.appDir, context.appPort, appOption)
      390 |       const res = await fetchViaHTTP(context.appPort, url)
    > 391 |       expect(res.status).toBe(200)
          |                          ^
      392 |       expect(res.headers.get('x-from-runtime')).toBe('false')
      393 |       expectNoError(moduleName)
      394 |     })

      at Object.toBe (integration/edge-runtime-module-errors/test/index.test.js:391:26)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge-unsupported-node | Change |
| - | - | - | - |
| buildDuration | 14s | 14s | N/A |
| buildDurationCached | 7.6s | 6.3s | N/A |
| nodeModulesSize | 198 MB | 198 MB | ✓ |
| nextStartRea..uration (ms) | 438ms | 435ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge-unsupported-node | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.7 kB | 30.7 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | ✓ |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 99.3 kB | 99.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge-unsupported-node | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge-unsupported-node | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge-unsupported-node | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge-unsupported-node | Change |
| - | - | - | - |
| index.html gzip | 529 B | 529 B | ✓ |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 525 B | 523 B | N/A |
| Overall change | 1.07 kB | 1.07 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge-unsupported-node | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.4 kB | 95.4 kB | N/A |
| page.js gzip | 3.04 kB | 3.04 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge-unsupported-node | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 624 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge-unsupported-node | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 96.9 kB | 96.9 kB | ✓ |
| app-page-tur..prod.js gzip | 98.6 kB | 98.6 kB | ✓ |
| app-page-tur..prod.js gzip | 93.1 kB | 93.1 kB | ✓ |
| app-page.run...dev.js gzip | 144 kB | 144 kB | ✓ |
| app-page.run..prod.js gzip | 91.6 kB | 91.6 kB | ✓ |
| app-route-ex...dev.js gzip | 21.4 kB | 21.4 kB | ✓ |
| app-route-ex..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 21 kB | 21 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.8 kB | 9.8 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-turbo...prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim...dev.js gzip | 23.1 kB | 23.1 kB | ✓ |
| pages.runtim..prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| server.runti..prod.js gzip | 50.7 kB | 50.7 kB | ✓ |
| Overall change | 945 kB | 945 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js edge-unsupported-node | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | N/A |
| index.pack gzip | 105 kB | 106 kB | ⚠️ +423 B |
| Overall change | 105 kB | 106 kB | ⚠️ +423 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: b3990ffd638ae87434a2db31a59ec58ad6f723c1

File: packages/next-swc/crates/next-core/src/next_import_map.rs [hunk 1]
```diff
@@ -38,6 +39,57 @@ use crate::{
     util::NextRuntime,
 };
 
+const NODE_INTERNALS: [&str; 48] = [
```
Comment (wbinnssmith) [inline packages/next-swc/crates/next-core/src/next_import_map.rs:42]: Could this be composed from https://github.com/vercel/turbo/blob/f714840e331fee4008b85a8d47687dc5c2a163f1/crates/turbopack-resolve/src/resolve.rs#L18-L71?

Comment (kwonoj) [reply packages/next-swc/crates/next-core/src/next_import_map.rs:42]: I was debating to leave comment about those - those 2 are actually different. next.js have some global augmentation (like buffer) and there are few apis available in runtime (https://edge-runtime.vercel.app/features/available-apis) .

File: packages/next-swc/crates/next-core/src/next_edge/unsupported.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            // packages/next/src/server/web/globals.ts augments global with
+            // `__import_unsupported` and necessary functions.
+            let code = formatdoc! {
+              r#"
+              __turbopack_export_namespace__(__import_unsupported(`{module}`));
+              "#
+            };
+            let content = AssetContent::file(File::from(code).into());
+            let source = VirtualSource::new(context, content);
+            return Ok(
```
Comment (wbinnssmith) [inline packages/next-swc/crates/next-core/src/next_edge/unsupported.rs:70]: Should this emit a `ResolvingIssue`[0] instead?

[0] https://github.com/vercel/turbo/blob/f714840e331fee4008b85a8d47687dc5c2a163f1/crates/turbopack-core/src/issue/resolve.rs#L18

Comment (kwonoj) [reply packages/next-swc/crates/next-core/src/next_edge/unsupported.rs:70]: I don't think so? if we hit this, this expects to correctly resolved virtual source and virtual source should raise runtime throw instead.

**Review by wbinnssmith — APPROVED** [review]

**Pull request merged**